### PR TITLE
feat(shared): #1632 create shared breadcrumb component

### LINKED ISSUE #1632: Move common layouts to separate generic components for reuse
Create a new issue for each one

- project header (logo, padlock, name / breadcrumb, actions)
-

### PR DESCRIPTION
**Pull request by webkhushboo**
closes #1632 #1633

### Notes

A summary of what was achieved in this PR

- [ ] Created shared breadcrumb component
- [ ] Added shared breadcrumb component inside monitor list component
- [ ] Added shared breadcrumb component inside token list component
- [ ] Added shared breadcrumb component inside rating component
- [ ] Added shared breadcrumb component inside ping list component 

image

### CODE REVIEW

File: web/src/app/shared/components/breadcrumb/breadcrumb.component.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<div fxLayout="row" fxLayout.xs="column" fxLayoutAlign="space-between center" fxLayoutGap="5px">
+  <span *ngIf="!isSmallScreen" class="avatar" matBadgeColor="accent" [matBadge]="project.repositories.length">
+    <img [src]="project.logoUrl" matBadgeSize="small" *ngIf="project.logoUrl; else imagePlaceHolder" />
+    <ng-template #imagePlaceHolder>
+      <mat-icon class="material-icons-outlined purple-color">dashboard</mat-icon>
+    </ng-template>
+  </span>
+  <mat-icon class="material-icons-outlined green-color">{{ typeIcon }}</mat-icon>
+  <a *ngFor="let item of breadCrumb" [routerLink]=[item.link] fxHide.lt-md>{{ item.title }} ></a>
+  <div>{{subTitle}}</div>
```
Comment (eddiejaoude) [inline web/src/app/shared/components/breadcrumb/breadcrumb.component.html]: Spacing

```suggestion
  <div>{{ subTitle }}</div>
```

File: web/src/app/shared/components/breadcrumb/breadcrumb.component.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+export class BreadcrumbComponent implements OnInit {
+
+  public typeIcon: string;
+
+  @Input() project: ProjectModel;
+  @Input() isSmallScreen: Boolean;
+  @Input() subTitle: string;
+  @Input() breadCrumb: BreadCrumbModel[];
+
+ngOnInit(): void {
```
Comment (eddiejaoude) [inline web/src/app/shared/components/breadcrumb/breadcrumb.component.ts]: Indentation doesn't look quite right

File: web/src/app/shared/shared.module.ts [hunk 3]
```diff
@@ -23,7 +24,8 @@ import { MarkdownDirective } from './directives/markdown.directive';
     DialogConfirmationComponent,
     DialogListComponent,
     ProjectsListComponent,
-    PrivatePublicProjectComponent],
+    PrivatePublicProjectComponent,
+    BreadcrumbComponent],
```
Comment (eddiejaoude) [inline web/src/app/shared/shared.module.ts]: Ideally the closing `]` should be on its own line - I realise it was like that before

File: web/src/app/monitors/monitors-list/monitors-list.component.ts [hunk 4]
```diff
@@ -66,10 +67,10 @@ export class MonitorsListComponent implements OnInit, OnDestroy {
       .subscribe((project: ProjectModel) => {
         this.project = project;
         this.monitors = project.monitors ? project.monitors : [];
+        this.breadCrumb = [{link: `/projects/${this.project.uid}`, title: this.project.title}];
```
Comment (eddiejaoude) [inline web/src/app/monitors/monitors-list/monitors-list.component.ts]: Spacing
```suggestion
        this.breadCrumb = [{ link: `/projects/${this.project.uid}`, title: this.project.title }];
```

File: web/src/app/monitors/pings-list/pings-list.component.ts [hunk 5]
```diff
@@ -46,7 +48,12 @@ export class PingsListComponent implements OnInit, OnDestroy {
   ngOnInit(): void {
     this.projectUid = this.route.snapshot.paramMap.get('projectUid');
     this.monitorUid = this.route.snapshot.paramMap.get('monitorUid');
-    this.route.data.subscribe((data: { pings: PingModel[] }) => this.pings = data.pings);
+    this.route.data.subscribe((data: { pings: PingModel[], project: ProjectModel }) => {
+      this.pings = data.pings;
+      this.project = data.project;
+      this.breadCrumb = [{ link: `/projects/${this.project.uid}`, title: this.project.title },
+      { link: `/projects/${this.project.uid}/monitors`, title: 'Monitors' }];
```
Comment (eddiejaoude) [inline web/src/app/monitors/pings-list/pings-list.component.ts]: I think this is more readable like


```typescript
this.breadCrumb = [
   { link: `/projects/${this.project.uid}`, title: this.project.title },
   { link: `/projects/${this.project.uid}/monitors`, title: 'Monitors' }
];
```

File: web/src/app/projects/rating/rating.component.ts [hunk 6]
```diff
@@ -28,9 +34,14 @@ export class RatingComponent implements OnInit, OnDestroy {
...
     private repositoryService: RepositoryService,
-    private route: ActivatedRoute
+    private route: ActivatedRoute,
+    private breakpointObserver: BreakpointObserver
   ) {
-    this.route.data.subscribe((data: { repository: RepositoryModel }) => this.repository = data.repository);
+    this.route.data.subscribe((data: { repository: RepositoryModel, project: ProjectModel }) => {
+      this.repository = data.repository;
+      this.project = data.project;
+      this.breadCrumb = [{link: `/projects/${this.project.uid}`, title: this.project.title}];
```
Comment (eddiejaoude) [inline web/src/app/projects/rating/rating.component.ts]: Spacing

```suggestion
      this.breadCrumb = [{ link: `/projects/${this.project.uid}`, title: this.project.title }];
```

**Comment by eddiejaoude** [comment]
Looks good. I left a few small comments.

Some thoughts:
- Should this be in the new component `this.breakpointObserver`, instead of in multiple the parents?
- I just realised when testing that create/edit `monitor` and `token` pages should have breadcrumb also
- `Project` view page should use the breadcrumb too

File: web/src/app/shared/components/breadcrumb/breadcrumb.component.ts (hunk 2 above)
Comment (webkhushboo) [reply web/src/app/shared/components/breadcrumb/breadcrumb.component.ts]: done

File: web/src/app/shared/shared.module.ts (hunk 3 above)
Comment (webkhushboo) [reply web/src/app/shared/shared.module.ts]: okay done

File: web/src/app/monitors/monitors-list/monitors-list.component.ts (hunk 4 above)
Comment (webkhushboo) [reply web/src/app/monitors/monitors-list/monitors-list.component.ts]: done

File: web/src/app/monitors/pings-list/pings-list.component.ts (hunk 5 above)
Comment (webkhushboo) [reply web/src/app/monitors/pings-list/pings-list.component.ts]: done

File: web/src/app/projects/rating/rating.component.ts (hunk 6 above)
Comment (webkhushboo) [reply web/src/app/projects/rating/rating.component.ts]: done

**Comment by webkhushboo** [comment]
> Looks good. I left a few small comments.
> 
> Some thoughts:
> 
> * Should this be in the new component `this.breakpointObserver`, instead of in multiple the parents?
> * I just realised when testing that create/edit `monitor` and `token` pages should have breadcrumb also
> * `Project` view page should use the breadcrumb too

While I was coding , I was also thinking to move `this.breakpointObserver` inside breadcrumb component but it was also used for table columns in the parent components so I didn't touched it.
I will add the `this.breakpointObserver` inside breadcrumb for `isSmallScreen` check.Let me know your thoughts on it.
I will add the breadcrumb in the create/edit `monitor` and `tokens` pages and view component too.

Question:
In view component we have menu for showing edit and delete actions for project in breadcrumb. Do we need to show this menu on all pages or just view component page ?

**Comment by eddiejaoude** [comment]
> While I was coding , I was also thinking to move this.breakpointObserver inside breadcrumb component but it was also used for table columns in the parent components so I didn't touched it.
I will add the this.breakpointObserver inside breadcrumb for isSmallScreen check.Let me know your thoughts on it.

Ah ok, if it is used for something else, then best to leave it where it is 👍 

> I will add the breadcrumb in the create/edit monitor and tokens pages and view component too.

Thank you 🤓 

> In view component we have menu for showing edit and delete actions for project in breadcrumb. Do we need to show this menu on all pages or just view component page ?

Good question! I think the easiest for now is to leave it on all page

File: web/src/app/shared/components/breadcrumb/breadcrumb.component.scss [hunk 7]
```diff
@@ -1,8 +1,13 @@
+div {
+  font-size: 20px;
+}
```
Comment (eddiejaoude) [inline web/src/app/shared/components/breadcrumb/breadcrumb.component.scss]: I think this should wrap `.avatar`?

**Comment by eddiejaoude** [comment]
Looks good. I left 1 inline comment.

I notice that the project logo/placeholder looks fine on the `dashboard` and `tokens` page, but on the `monitors` page it still has the Dashboardhub placeholder (screenshots below)

## Dashboard page

Screenshot 2019-12-28 10 43 14

## Tokens page

Screenshot 2019-12-28 10 43 36

## Monitors page

Screenshot 2019-12-28 10 43 23

File: web/src/app/shared/components/breadcrumb/breadcrumb.component.scss (hunk 7 above)
Comment (webkhushboo) [reply web/src/app/shared/components/breadcrumb/breadcrumb.component.scss]: okay

**Comment by webkhushboo** [comment]
> Looks good. I left 1 inline comment.
> 
> I notice that the project logo/placeholder looks fine on the `dashboard` and `tokens` page, but on the `monitors` page it still has the Dashboardhub placeholder (screenshots below)
> 
Yes you are right , we are showing fallback icon as dashboard hub logo in monitor list component. I have removed it now.
> ## Dashboard page
> Screenshot 2019-12-28 10 43 14
> 
> ## Tokens page
> Screenshot 2019-12-28 10 43 36
> 
> ## Monitors page
> Screenshot 2019-12-28 10 43 23

**Review by eddiejaoude — APPROVED** [review]

**Pull request merged**